apiserver/remoterelations/remoterelations.go

```diff
@@ -184,55 +184,134 @@ func (api *RemoteRelationsAPI) RelationUnitSettings(relationUnits params.Relatio
 	return results, nil
 }
 
-// Relations returns information about the cross-model relations with the specified keys
-// in the local model.
-func (api *RemoteRelationsAPI) Relations(entities params.Entities) (params.RemoteRelationResults, error) {
-	results := params.RemoteRelationResults{
-		Results: make([]params.RemoteRelationResult, len(entities.Entities)),
+// IngressNetworksForRelation returns any CIDRs for which ingress is required to allow
+// the specified relations to properly function.
+func (api *RemoteRelationsAPI) IngressNetworksForRelation(entities params.Entities) (params.IngressNetworkResults, error) {
+	results := params.IngressNetworkResults{
+		Results: make([]params.IngressNetworkResult, len(entities.Entities)),
 	}
-	one := func(entity params.Entity) (*params.RemoteRelation, error) {
-		tag, err := names.ParseRelationTag(entity.Tag)
+	one := func(relTag params.Entity) (*params.IngressNetworkInfo, error) {
+		// Load the relation details for the current tag.
+		remoteRelation, err := api.remoteRelation(relTag)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		rel, err := api.st.KeyRelation(tag.Id())
+
+		// Double check we have a source model - we should never hit this.
+		if remoteRelation.SourceModelUUID == "" {
+			return nil, errors.NotValidf("remote relation %v with empty source model UUID", remoteRelation)
+		}
+
+		var result params.IngressNetworkInfo
+		result.ApplicationName = remoteRelation.ApplicationName
+
+		// Until networking support becomes more sophisticated, for now
+		// we only care about opening access to applications with an endpoint
+		// having the "provider" role. The assumption is that such endpoints listen
+		// to incoming connections and thus require ingress. An exception to this
+		// would be applications which accept connections onto an endpoint which
+		// has a "requirer" role.
+		if remoteRelation.Endpoint.Role != charm.RoleProvider {
+			logger.Debugf(
+				"no ingress network for application %v without provides endpoint %v",
+				remoteRelation.ApplicationName, remoteRelation.Endpoint.Name)
+			return &result, nil
+		}
+
+		// Get the source model of the remote application so we can determine the
+		// required CIDRs for which we need to allow ingress to this model.
+		sourceSt, closer, err := api.pool.Get(remoteRelation.SourceModelUUID)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		result := &params.RemoteRelation{
-			Id:   rel.Id(),
-			Life: params.Life(rel.Life().String()),
-			Key:  tag.Id(),
+		defer closer()
+
+		// TODO(wallyworld) - as a first implementation, just get all CIDRs from the model
+		all, err := sourceSt.AllIPAddresses()
```

Comment:
I tried AllSubnets but that collection wasn't populated. From what I could see in the AWS provider code, the NetworkInterface implementation ends up setting the IPAddresses collection. I could see CLI which updated subnets, but not anything from the provider, and this seemed to match the observed behaviour. Maybe I'm missing something. So I only used IPAddresses() because that seemed to work.